src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,468 @@
+
+        // Now we shape the angle by applying the circles.Note that only one of them will have an
+        // effect, based on the radius that we modified set through `is_inwards`.
+        let sculpted_shape = base_shape;
+        let sculpted_shape = Intersection(sculpted_shape,circle_outer);
+        let sculpted_shape = Difference(sculpted_shape,circle_inner);
+        let sculpted_shape = sculpted_shape.fill(Srgb::new(0.26, 0.69, 0.99));
+
+        // The angle should be centered on (0,0) to make it easier to rotate and minimise the
+        // required canvas.
+        let center_offset    = Var::<Distance<Pixels>>::from(&shape_height * Var::from(0.5));
+        let sculpted_shape   = sculpted_shape.translate_y(-center_offset);
+
+        // This is a conditional rotation that allows the port to either point inwards
+        // or outwards.
+        let pi             = Var::from(PI);
+        let rotation_angle = is_inwards * pi;
+        let rotation_angle = Var::<Angle<Radians>>::from(rotation_angle);
+        let sculpted_shape = sculpted_shape.rotate(rotation_angle);
+
+        sculpted_shape.into()
+    }
+
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32,is_inwards:f32) {
+            // NOTE: `is_inwards` should only be 0.0 or 1.0.
+            new_port(height,width,inner_radius,is_inwards)
+        }
+    }
+
+    impl Shape{
+        /// Set the shape parameters derived from the `Specification`.
+        pub fn update_from_spec(&self,spec:&Specification){
+            self.height.set(spec.height);
+            self.inner_radius.set(spec.inner_radius);
+            self.width.set(spec.width.value.to_radians());
+            match &spec.direction{
+                Direction::In  => self.is_inwards.set(1.0),
+                Direction::Out => self.is_inwards.set(0.0),
+            };
+        }
+    }
+}
+
+
+
+// ============
+// === Port ===
+// ============
+
+/// Shape view for input port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for output port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
```

Comment:
These 2.0 * etc are OK, but are not described anywhere. You should refactor `2.0` to a variable and name it in such way that it will explain to the reader what it is really used for (it is the scaling factor to make place for growing shapes when zooming out, right?)